lib/rubocop/cop/mixin/surrounding_space.rb

```diff
@@ -107,9 +107,9 @@ def empty_offense(node, range, message, command)
         end
       end
 
-      def empty_brackets?(left_bracket_token, right_bracket_token)
-        left_index = processed_source.tokens.index(left_bracket_token)
-        right_index = processed_source.tokens.index(right_bracket_token)
+      def empty_brackets?(tokens, left_bracket_token, right_bracket_token)
+        left_index = tokens.index(left_bracket_token)
+        right_index = tokens.index(right_bracket_token)
```

Comment:
`processed_source` and `tokens` are attribute accesses:

- https://github.com/rubocop/rubocop/blob/v1.43.0/lib/rubocop/cop/base.rb#L43
- https://github.com/rubocop/rubocop-ast/blob/v1.24.1/lib/rubocop/ast/processed_source.rb#L18

I don't understand why this change makes it so much faster yet. AFAIK, unlike something like benchmark-ips, this approach using `bin/rubocop-profile` some changes the performance metrics and ranking (especially low ranking) .

Reply:
Ah, maybe I understand. It mean that pre-filtering with `tokens_within` is faster, right?

OTOH, compatibility due to signature change is a concern you're pointing out. It doesn't seem to be used in the current RuboCop HQ repos, but 3rd party custom cops are unknown. It may be polite to create compatibility, but may it be a waste of complexity. It's a little difficult to judge.

Reply:
@koic 

> Ah, maybe I understand. It mean that pre-filtering with tokens_within is faster, right?

Yes, exactly! In most cases we already filter a subset of tokens and can simply pass them to `empty_brackets?`. I'll added this note to the PR description. Sorry for missing this previously :pray: 

> but 3rd party custom cops are unknown.

Yes, I was concerned about this. On the other I am also worried about additional (and perhaps unneeded) complexity to maintain backwards-compatibility.

Would it work to change the signature now, add a changelog entry, and only adjust this signature (making it backward-compatible) after users complain? :thinking:

Reply:
I wish we were able to pass offset to `Array#index` (similar to `String#index`). See https://bugs.ruby-lang.org/issues/17056 :pray:

Reply:
@koic 

> It may be polite to create compatibility, but may it be a waste of complexity. It's a little difficult to judge.

If changing the method signature is a concern and potentially slowing down the merge of this PR perhaps we could make it backwards-compatible like:

```ruby
      def empty_brackets?(left_bracket_token, right_bracket_token, tokens: processed_source.tokens)
```

WDYT?

Reply:
@koic 

Done :heavy_check_mark: in a separate commit. I can revert it if you dislike the idea.